File: sdks/go/stubs.go
package crabka

import "context"

type Database struct{ client *Client }

func (d *Database) Connect(context.Context, string) error {
	return gatedError("database", "chapter-f-control-plane")
}

type Auth struct{ client *Client }

func (a *Auth) BearerToken() string {
	return a.client.bearerToken
}

func (a *Auth) SignIn(context.Context, string, string) error {
	return errorWithMessage(Unauthenticated, "identity APIs are not part of contract v1")
}

type Blob struct{ client *Client }

func (b *Blob) Put(context.Context, string, []byte) error {
	return gatedError("blob", "chapter-b-blob-api")
}

func (b *Blob) Get(context.Context, string) ([]byte, error) {
	return nil, gatedError("blob", "chapter-b-blob-api")
}

File: sdks/ts/src/stubs.ts
import { UnauthenticatedError, UnimplementedError } from "./errors.ts";

export const QUEUES_GATE = "gateway-sharegroup-rpc";
export const DATABASE_GATE = "chapter-f-control-plane";
export const BLOB_GATE = "chapter-b-blob-api";

export type DatabaseModule = {
  connect(name: string): Promise<never>;
};

export type AuthModule = {
  bearerToken?: string;
  signIn(username: string, password: string): Promise<never>;
};

export type BlobModule = {
  put(key: string, value: Uint8Array): Promise<never>;
  get(key: string): Promise<never>;
};

export function createDatabaseModule(): DatabaseModule {
  return {
    async connect() {
      throw new UnimplementedError("database", DATABASE_GATE);
    },
  };
}

export function createAuthModule(bearerToken?: string): AuthModule {
  return {
    bearerToken,
    async signIn() {
      throw new UnauthenticatedError("identity APIs are not part of contract v1");
    },
  };
}

export function createBlobModule(): BlobModule {
  return {
    async put() {
      throw new UnimplementedError("blob", BLOB_GATE);
    },
    async get() {
      throw new UnimplementedError("blob", BLOB_GATE);
    },
  };
}
